### PR TITLE
[WIP] [KYUUBI #825] Introduce jcommander framework to kyuubi project

### LINKED ISSUE #825: Introduce jcommander framework to kyuubi project
### 1. Describe the feature

sub task from issue #775 

Introduce jcommander framework to kyuubi project

### PR DESCRIPTION
**Pull request by timothy65535**
### _Why are the changes needed?_

Fixes #825

For more details, please go to #775 

Hi @yaooqinn @pan3793, 

Here's the patch base on JCommander in scala, please point out if the scala grammar is not used correctly.

The following usage is generated from JCommander, I think it seems a bit long for users, we can discuss it here.
By the way, we can use the format of bookkeeper-ctl‘s usage.

Thanks.
```
Usage: kyuubi-ctl <server|engine> <create|get|delete|list> [options]
[ ENGINE ]
get
  -s, --host       Hostname or IP address of a service.
  -u, --user       The user name this engine belong to.
  -n, --namespace  The namespace, using kyuubi-defaults/conf if absent.
  -p, --port       Listening port of a service.
  -zk, --zk-quorum The connection string for the zookeeper ensemble, using zk quorum manually.
  -h, --help       
delete
  -zk, --zk-quorum The connection string for the zookeeper ensemble, using zk quorum manually.
  -n, --namespace  The namespace, using kyuubi-defaults/conf if absent.
  -p, --port       Listening port of a service.
  -s, --host       Hostname or IP address of a service.
  -u, --user       The user name this engine belong to.
  -h, --help       
list
  -zk, --zk-quorum The connection string for the zookeeper ensemble, using zk quorum manually.
  -p, --port       Listening port of a service.
  -n, --namespace  The namespace, using kyuubi-defaults/conf if absent.
  -u, --user       The user name this engine belong to.
  -s, --host       Hostname or IP address of a service.
  -h, --help       

[ SERVER ]
create
  -s, --host       Hostname or IP address of a service.
  -n, --namespace  The namespace, using kyuubi-defaults/conf if absent.
  -zk, --zk-quorum The connection string for the zookeeper ensemble, using zk quorum manually.
  -p, --port       Listening port of a service.
  -h, --help       
get
  -n, --namespace  The namespace, using kyuubi-defaults/conf if absent.
  -s, --host       Hostname or IP address of a service.
  -p, --port       Listening port of a service.
  -zk, --zk-quorum The connection string for the zookeeper ensemble, using zk quorum manually.
  -h, --help       
delete
  -s, --host       Hostname or IP address of a service.
  -n, --namespace  The namespace, using kyuubi-defaults/conf if absent.
  -zk, --zk-quorum The connection string for the zookeeper ensemble, using zk quorum manually.
  -p, --port       Listening port of a service.
  -h, --help       
list
  -p, --port       Listening port of a service.
  -zk, --zk-quorum The connection string for the zookeeper ensemble, using zk quorum manually.
  -s, --host       Hostname or IP address of a service.
  -n, --namespace  The namespace, using kyuubi-defaults/conf if absent.
  -h, --help  
```

#### 1. Totoal Usage
```
[bookkeeper-all-4.13.0]$ bin/bkctl
bkctl interacts and operates Apache BookKeeper clusters

Usage:  bkctl [flags] [command group] [commands]

    cookie              Commands on operating cookies

Infrastructure commands :

    autorecovery        Command on some specific operation.
    bookie              Commands on operating a single bookie
    bookieid            Commands operating on bookie ids
    bookies             Commands on operating a cluster of bookies
    cluster             Commands on administrating bookkeeper clusters
```

#### 2. When choose one command
````
[bookkeeper-all-4.13.0]$ bin/bkctl bookie
Commands on operating a single bookie

Usage:  bkctl bookie [command] [command options]

Commands:

    converttodbstorage                              Convert bookie indexes from
                                                    InterleavedStorage to DbLedgerStorage format
    converttointerleavedstorage                     Convert bookie indexes from
                                                    DbLedgerStorage to InterleavedStorage
                                                    format
    endpointinfo                                    Get all end point
                                                    information about a given bookie.
    flip-bookie-id                                  Update bookie id in ledgers
                                                    (this may take a long time).
    format                                          Format the current server

````

### _How was this patch tested?_
- [X] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/common/AbstractCommand.scala [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.ctl.commands.common
+
+import com.beust.jcommander.JCommander
+
+trait AbstractCommand {
+
+  def run(jc: JCommander): Unit = {}
```
Comment (yaooqinn) [inline kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/common/AbstractCommand.scala:24]: Can we have possible output for each command？

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  }
+
+}
+
+object KyuubiCli extends Logging {
+
+  def main(args: Array[String]): Unit = {
+
+    val shell = new KyuubiCli
+    if (args(0).equals("--help")) {
```
Comment (yaooqinn) [inline kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala:99]: this is hardcoded, uses shall get help whenever an error occurs or they request help explictly

File: kyuubi-ctl/pom.xml [hunk 3]
```diff
@@ -109,6 +109,11 @@
             <version>${project.version}</version>
             <scope>test</scope>
         </dependency>
+
+        <dependency>
+            <groupId>com.beust</groupId>
+            <artifactId>jcommander</artifactId>
+        </dependency>
```
Comment (pan3793) [inline kyuubi-ctl/pom.xml:116]: put compile scope deps before thoes test scope

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/engine/ListEngineCommand.scala [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  @Parameter(
+    names = Array("-n", "--namespace"),
+    description = "The namespace, using kyuubi-defaults/conf if absent.",
+    order = 1)
+  var namespace: String = _
+
+  @Parameter(
+    names = Array("-s", "--host"),
+    description = "Hostname or IP address of a service.",
+    order = 1)
```
Comment (pan3793) [inline kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/engine/ListEngineCommand.scala:42]: all parameters use same`order` value?

**Comment by pan3793** [comment]
After second thought and have a discussion with @turboFei (author of this the first version of `kyuubi-ctl`),  I change my vote to `kyuubi-ctl <create|get|delete|list> <server|engine> options`. It's because
1. some tools use `cmd action object`, i.e. `kubectl`, some tools chose `cmd subcmd subcmd`, both are widely used.
2. one of the fact is that `kyuubi-ctl list service` and `kyuubi-ctl list engine` are almost the same things except for the `znode root path`, and we can reuse lots code if chose the pattern `cmd action object`

**Comment by yaooqinn** [comment]
I am planning that we have a command like `kyuubi-ctl [create|get|delete|list] config` or `kyuubi-ctl config [create|get|delete|list]`. with this, which way is right to go? @turboFei @timothy65535 @pan3793

**Comment by pan3793** [comment]
@yaooqinn would you please add some details or demos on `config`? in my view, the chosen will be based on the code reused rate.

**Comment by yaooqinn** [comment]
- create, apply a config to a running server(maybe)or a configuration file
 - list, get and validate all configurations in the configuration file(can be configured with -f -type (spark|kyuubi))
 - get, get a single config's config entry with doc, version, key, value, etc
 - delete, unapply a config to a running server(maybe) or a configuration file
 
or something.

**Comment by pan3793** [comment]
`kyuubi-ctl config [create|get|delete|list]` makes more sense to me, and I suggest change `create` to `set`, `delete` to `remove` for object `config`, in other words, we can use independent action words for different objects.

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/common/AbstractCommand.scala (hunk 1 above)
Comment (timothy65535) [reply kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/common/AbstractCommand.scala:24]: I'm sorry, I didn't get your points.

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala (hunk 2 above)
Comment (timothy65535) [reply kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala:99]: right, I'm thinking how to improve it.

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+      .addCommand(new DeleteServerCommand)
+      .addCommand(new ListServerCommand)
+      .build()
+
+    // register commands
+    HashMap(
+      ServiceType.ENGINE -> engineCommand,
+      ServiceType.SERVER -> serverCommand
+    )
+  }
```
[the remark this answers is not in the document]
Comment (timothy65535) [reply kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala:63]: Got.

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/engine/ListEngineCommand.scala (hunk 4 above)
Comment (timothy65535) [reply kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/engine/ListEngineCommand.scala:42]: Good catch.

**Comment by timothy65535** [comment]
Hi @yaooqinn @pan3793, review again when you are free , thanks.

### cmd examples
1. input `--help`
```
Usage: kyuubi-ctl <server|engine|config> [actions] [options] 

  engine  Commands on operating engine
  server  Commands on operating server
  config  Commands on operating config
```

2. input `server --help`
```
Usage: kyuubi-ctl server [actions] [options]
create
  -zk, --zk-quorum The connection string for the zookeeper ensemble, using zk quorum manually.
  -n, --namespace  The namespace, using kyuubi-defaults/conf if absent.
  -s, --host       Hostname or IP address of a service.
  -p, --port       Listening port of a service.
  -h, --help       
get
  -zk, --zk-quorum The connection string for the zookeeper ensemble, using zk quorum manually.
  -n, --namespace  The namespace, using kyuubi-defaults/conf if absent.
  -s, --host       Hostname or IP address of a service.
  -p, --port       Listening port of a service.
  -h, --help       
delete
  -zk, --zk-quorum The connection string for the zookeeper ensemble, using zk quorum manually.
  -n, --namespace  The namespace, using kyuubi-defaults/conf if absent.
  -s, --host       Hostname or IP address of a service.
  -p, --port       Listening port of a service.
  -h, --help       
list
  -zk, --zk-quorum The connection string for the zookeeper ensemble, using zk quorum manually.
  -n, --namespace  The namespace, using kyuubi-defaults/conf if absent.
  -s, --host       Hostname or IP address of a service.
  -p, --port       Listening port of a service.
  -h, --help       
```

3. input `server get --help`
```
get
  -zk, --zk-quorum The connection string for the zookeeper ensemble, using 
                     zk quorum manually.
  -n, --namespace  The namespace, using kyuubi-defaults/conf if absent.
  -s, --host       Hostname or IP address of a service.
  -p, --port       Listening port of a service.
  -h, --help 
```

### Easy to add new more service / command
1. Add new command which extends `Command`
2. Add the new command to service command group
3. If add new service type, need to update `Kyuubi`
4. ok

image

File: kyuubi-ctl/pom.xml (hunk 3 above)
Comment (timothy65535) [reply kyuubi-ctl/pom.xml]: done.

**Comment by yaooqinn** [comment]
cc @turboFei PTAL

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/common/UnixStyleUsage.scala [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.List
+
+import scala.collection.JavaConverters._
+
+import com.beust.jcommander.{DefaultUsageFormatter, JCommander, ParameterDescription}
+
+class UnixStyleUsage(commander: JCommander) extends DefaultUsageFormatter(commander) {
+
+  override def appendMainLine(out: Builder, hasOptions: Boolean,
+                              hasCommands: Boolean, indentCount: Int, indent: String): Unit = {
```
Comment (yaooqinn) [inline kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/common/UnixStyleUsage.scala:30]:
```suggestion
          hasCommands: Boolean, indentCount: Int, indent: String): Unit = {
```

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/common/UnixStyleUsage.scala [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    wrapDescription(out, indentCount, commander.getProgramDisplayName + "\n")
+  }
+
+  override def appendAllParametersDetails(out: Builder, indentCount: Int, indent: String,
+                                          sortedParameters: List[ParameterDescription]): Unit = {
```
Comment (yaooqinn) [inline kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/common/UnixStyleUsage.scala:35]:
```suggestion
  override def appendAllParametersDetails(out: Builder, indentCount: Int, indent: String,
      sortedParameters: List[ParameterDescription]): Unit = {
```

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/engine/DeleteEngineCommand.scala [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+  @Parameter(names = Array("-h", "--help"), help = true, order = 5)
+  var help: Boolean = false
+
+  override def run(jc: JCommander): Unit = {
+    if (help) {
+      jc.setUsageFormatter(new UnixStyleUsage(jc))
+      jc.usage()
+    } else {
+      // TODO
```
Comment (yaooqinn) [inline kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/engine/DeleteEngineCommand.scala:65]: what is this for?

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/engine/GetEngineCommand.scala [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    if (help) {
+      jc.setUsageFormatter(new UnixStyleUsage(jc))
+      jc.usage()
+    } else {
+      // TODO
+    }
+  }
+
+  @VisibleForTesting
+  def test(): (String, String, Int) = {
```
Comment (yaooqinn) [inline kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/engine/GetEngineCommand.scala:71]: is it necessary?

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/server/CreateServerCommand.scala [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+  @Parameter(names = Array("-h", "--help"), help = true, order = 5)
+  var help: Boolean = false
+
+  override def run(jc: JCommander): Unit = {
+    if (help) {
+      jc.setUsageFormatter(new UnixStyleUsage(jc))
+      jc.usage()
+    } else {
+      // TODO
```
Comment (yaooqinn) [inline kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/server/CreateServerCommand.scala:59]: ditto

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/engine/DeleteEngineCommand.scala (hunk 8 above)
Comment (timothy65535) [reply kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/engine/DeleteEngineCommand.scala:65]: `TODO` main logic will be put here.

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/engine/GetEngineCommand.scala (hunk 9 above)
Comment (timothy65535) [reply kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/engine/GetEngineCommand.scala:71]: yes. it's used for testing.

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/server/CreateServerCommand.scala (hunk 10 above)
Comment (timothy65535) [reply kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/server/CreateServerCommand.scala:59]: the logic for the command will put here.

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/engine/GetEngineCommand.scala (hunk 9 above)
Comment (yaooqinn) [reply kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/engine/GetEngineCommand.scala:71]: the attributes are public

Comment (timothy65535) [reply kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/engine/GetEngineCommand.scala:71]: These commands did not return value, so need this.

https://github.com/apache/bookkeeper/tree/master/tools/ledger/src/test/java/org/apache/bookkeeper/tools/cli/commands/bookie

**Comment by yaooqinn** [comment]
> Hi @yaooqinn @pan3793 @turboFei, review again when you are free , thanks.
> 
> ### cmd examples
> 1. input `--help`
> 
> ```
> Usage: kyuubi-ctl <server|engine|config> [actions] [options] 
> 
>   engine  Commands on operating engine
>   server  Commands on operating server
>   config  Commands on operating config
> ```
> 
> 1. input `server --help`
> 
> ```
> Usage: kyuubi-ctl server [actions] [options]
> create
>   -zk, --zk-quorum The connection string for the zookeeper ensemble, using zk quorum manually.
>   -n, --namespace  The namespace, using kyuubi-defaults/conf if absent.
>   -s, --host       Hostname or IP address of a service.
>   -p, --port       Listening port of a service.
>   -h, --help       
> get
>   -zk, --zk-quorum The connection string for the zookeeper ensemble, using zk quorum manually.
>   -n, --namespace  The namespace, using kyuubi-defaults/conf if absent.
>   -s, --host       Hostname or IP address of a service.
>   -p, --port       Listening port of a service.
>   -h, --help       
> delete
>   -zk, --zk-quorum The connection string for the zookeeper ensemble, using zk quorum manually.
>   -n, --namespace  The namespace, using kyuubi-defaults/conf if absent.
>   -s, --host       Hostname or IP address of a service.
>   -p, --port       Listening port of a service.
>   -h, --help       
> list
>   -zk, --zk-quorum The connection string for the zookeeper ensemble, using zk quorum manually.
>   -n, --namespace  The namespace, using kyuubi-defaults/conf if absent.
>   -s, --host       Hostname or IP address of a service.
>   -p, --port       Listening port of a service.
>   -h, --help       
> ```
> 
> 1. input `server get --help`
> 
> ```
> get
>   -zk, --zk-quorum The connection string for the zookeeper ensemble, using 
>                      zk quorum manually.
>   -n, --namespace  The namespace, using kyuubi-defaults/conf if absent.
>   -s, --host       Hostname or IP address of a service.
>   -p, --port       Listening port of a service.
>   -h, --help 
> ```
> 
> ### Easy to add new more service / command
> 1. Add new command which extends `Command`
> 2. Add the new command to service command group
> 3. If add new service type, need to update `Kyuubi`
> 4. ok
> 
> image

the behavior changes a lot for each command, the acceptable args change from cmd to cmd

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/engine/GetEngineCommand.scala (hunk 9 above)
Comment (timothy65535) [reply kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/engine/GetEngineCommand.scala]: Got, done.

**Comment by timothy65535** [comment]
The usage in master branch 
```
Usage: kyuubi-ctl [create|get|delete|list] [options]

  -zk, --zk-quorum <value>
                           The connection string for the zookeeper ensemble, using zk quorum manually.
  -n, --namespace <value>  The namespace, using kyuubi-defaults/conf if absent.
  -s, --host <value>       Hostname or IP address of a service.
  -p, --port <value>       Listening port of a service.
  -v, --version <value>    Using the compiled KYUUBI_VERSION default, change it if the active service is running in another.
  -b, --verbose            Print additional debug output.

Command: create [server]

Command: create server
        Expose Kyuubi server instance to another domain.

Command: get [server|engine] [options]
        Get the service/engine node info, host and port needed.
Command: get server
        Get Kyuubi server info of domain
Command: get engine
        Get Kyuubi engine info belong to a user.
  -u, --user <value>       The user name this engine belong to.

Command: delete [server|engine] [options]
        Delete the specified service/engine node, host and port needed.
Command: delete server
        Delete the specified service node for a domain
Command: delete engine
        Delete the specified engine node for user.
  -u, --user <value>       The user name this engine belong to.

Command: list [server|engine] [options]
        List all the service/engine nodes for a particular domain.
Command: list server
        List all the service nodes for a particular domain
Command: list engine
        List all the engine nodes for a user
  -u, --user <value>       The user name this engine belong to.

  -h, --help               Show help message and exit.

```

**Comment by timothy65535** [comment]
Each different command uses its own command parameters to be more extensible.

File: kyuubi-ctl/pom.xml [hunk 11]
```diff
@@ -109,6 +114,7 @@
             <version>${project.version}</version>
             <scope>test</scope>
         </dependency>
+
```
Comment (pan3793) [inline kyuubi-ctl/pom.xml:117]: Unnecessary change

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    ServiceType.engine -> new EngineCommandGroup,
+    ServiceType.server -> new ServerCommandGroup,
+    ServiceType.config -> new ConfigCommandGroup
+  )
+
+  def printAllCommandUsage(): Unit = {
+    // scalastyle:off println
+    val sj = new StringJoiner("|", "<", ">")
+    ServiceType.values.foreach(v => sj.add(String.valueOf(v)))
+    println(s"Usage: kyuubi-ctl ${sj.toString} [actions] [options] \n")
```
Comment (pan3793) [inline kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala:45]: `${sj.toString}` => `$sj`

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+}
+
+object KyuubiCli extends Logging {
+
+  def main(args: Array[String]): Unit = {
+    val shell = new KyuubiCli
+
+    if (args == null || args(0).equals("-h") || args(0).equals("--help") || args.length == 1) {
```
Comment (pan3793) [inline kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala:114]: is it feasible to leverage JCommander parser the top-level args? if does, let's also support `-v` `--version`

Comment (pan3793) [reply kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala:114]: maybe `args == null` => `args.isEmpty`?

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+      shell.printAllCommandUsage()
+    } else {
+      try {
+        shell.run(ServiceType.withName(args(0)), args.drop(1))
+      } catch {
+        case _: NoSuchElementException =>
+          error(s"unknown service type: args(0)")
```
Comment (pan3793) [inline kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala:121]: `args(0)` => `${args.head}`

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/common/UnixStyleUsage.scala [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.ctl.commands.common
+
+import java.lang.{StringBuilder => Builder}
```
Comment (pan3793) [inline kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/common/UnixStyleUsage.scala:20]: why rename?

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala (hunk 12 above)
Comment (timothy65535) [reply kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala:45]: `ServiceType.values.mkString("|", "<", ">")` result is `|server<engine<config>`

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/common/UnixStyleUsage.scala (hunk 15 above)
Comment (timothy65535) [reply kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/commands/common/UnixStyleUsage.scala:20]: the scala StringBuilder is not compatible with java.lang.StringBuilder.

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala (hunk 13 above)
Comment (timothy65535) [reply kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala:114]: `-v` `--version` is option for each command? 

It seems the `-v` param is never used in original command.
image

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala (hunk 14 above)
Comment (timothy65535) [reply kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/KyuubiCli.scala:121]: Good catch

**Comment by timothy65535** [comment]
Hi @pan3793 @yaooqinn thanks for the suggestion, patch is updated accordingly, can you please take another look. BTW, sorry for the delay, there are a lot of things to deal with in the past week.

**Comment by yaooqinn** [comment]
maybe we can do it group by group

**Comment by timothy65535** [comment]
> maybe we can do it group by group

First step is introduce framework to kyuubi project, we can discuss the design of the command frammework.

Next plan:
- [ ] Implement engine part
- [ ] Implement service part
